### DISCUSSION TITLE
Layout for the punctuation signs

### DISCUSSION THREAD

**Discussion by Lobo-Feroz** [binarybottle/engram-es#30] [General]
Hello there,

We're in the verge of having the first functional layout of engram-es. After toying a bit with it in type-fu, it feels very pleasant, and I very much would like to be able to start using it. 

We're missing the layout for the punctuation signs in the two center columns. Since @binarybottle 's algorithm does not analyze this part (potential room for further development maybe?) at this point we need to resort to manual design. 

I've added up some frequencies for the symbols here:

[spanish-character-frequency-v1 - signs - 20200821-01.ods](https://github.com/binarybottle/engram-es/files/7025495/spanish-character-frequency-v1.-.signs.-.20200821-01.ods)

Here's a sample but notice that the first 2 columns, which are the symbol frequencies, have more rows than shown here:

![signs - frequencies - 20200821 - key grouping](https://user-images.githubusercontent.com/10242383/130321797-ba4c21f9-26ee-454f-9269-29f6518c26d0.png)

First 2 columns are the aggregate frequency for a symbol (I've grouped the accute accents, the variants for the quotes and the apostrophe and so on).

Second set of columns are aggregates for _potential_ keys. Notice that they are only potentials, this is just a proposal and nothing is set in stone.

Two more design criteria: 

- I don't have data on which keys are easier, though obviously home row is best, and I suppose the upper row is easier than the bottom row for the center columns.
- engram-en has a nice simmetry for opening-closing symbols ([engram-en layout](https://raw.githubusercontent.com/binarybottle/engram/master/assets/engram-800px.png)). I'd like to keep that. It also shows this is part of the engram family.

First, I thought about having comma and period in the same key. In fact, linguistically they have the same function: a short stop and a longer stop. Period being the shifted version of the comma. Having this, we can set the acute accent and the comma-period key in the home row. As you can see, the summed frequencies for these symbols is 3.9 which is much more than the rest of the symbols combined (not counting that we could also have the other diacritics ( ¨ ` ^ ) here, but these are very low frequency anyway).
 
Also, I think it's best if the accute accent goes in the right hand because it's mostly used with vowels which are in the left hand. This puts the comma-period in the left hand, which as Ian has pointed out, will cause some SFB as many words end with ( e a ). But also many other words end with ( s n r ) which are in the right hand, so we save some SFB here.

The base layout for these symbols would be this:

![imagen](https://user-images.githubusercontent.com/10242383/130322090-386cb0af-6b80-4fc0-bc25-936efddf670d.png)

Yesterday I tested this design and it seems easy on the fingers. So let's see what we can do from here, based on the frequencies.

We have a bunch of symbols remaining. The hyphen (and the underscore) is the next more frequent one. And from there the frequencies start being lower and more similar. We have to place here:

- hyphen and underscore ( - _ ) (1 key) 
- opening and closing parenthesis ( ( ) ) (2 keys)
- opening and closing interrogations and exclamations ( ¿ ¡ ) ( ? ! ) (2 keys)
- colon and semicolon ( ; : ) (1 key)
- quotes and apostrophes ( " ' ) (1 key)

Too many keys for four spaces! So I thought about doing some more grouping. For instance, we could place the parenthesis with the interrogation and exclamation. And for instance, we can place them in the bottom row (where they are in engram-en):

![layout - signs - comma - accent - interrogation](https://user-images.githubusercontent.com/10242383/130322778-29b30e01-d25e-4551-ab0f-180c2d4f5855.png)

Now we're down to two empty keys and we still have to place ( - _ ) ( ; : ) ( " ' ). One has to go elsewhere.

This is where I need your help. I'd like to know how hard/easy compared to the others is this ISO key:

![imagen](https://user-images.githubusercontent.com/10242383/130322387-40178e98-77e5-46a1-883b-fdc5c0bb3a3d.png)

Subjectively it feels easy to me since it's moving the little finger a little way down. But it might not be the same for everybody and, of course, this key is missing in ANSI keyboards. But we could place something useful here.

Another thing we could do is place the ( ; : ) in the same comma-period key. AltGr + , = ;  Shift+AltGr + , = :
But I think this is quite complex and these symbols are somewhat used. But maybe, having these here could be an alternative for those who have ANSI keyboards (since the majority of spanish writers will use ISO anyway).

That way we could have ( - _ ) and ( " ' ) in the upper row, ( ; : ) in the pinky ISO key, and  ( ; : ) also as alt versions of the ( , . ) for ANSI users.

That would yield this:

![imagen](https://user-images.githubusercontent.com/10242383/130322606-5aec1fbe-350e-459a-8562-4aeccd9cd78e.png)

In the upper row I've put hyphen on the left and quotes on the right since the left key seems more accessible in staggered keyboards (thus it goes to the hyphen which has more frequency), and also to keep the quotes in the right since engram-en has them there. But these could be switched.

I think this could be functional and a viable layout. The only think I'm not liking too much is having ( ¡ ! ) in the AltGr layer instead of the shift layer, but we could also have them in the diacritics dead key layer for these keys. I'd rather have the exclamations in the shift layer and have the parenthesis elsewere (but still in a comfortable position, since the parenthesis have high frequencies). I'll think a bit about that.

What do you guys think?

**Comment by binarybottle** [comment]
Thank you, @Lobo-Feroz!
I was planning on tackling punctuation today, so this is a helpful start!

**Comment by NickG13** [comment]
Hi Miguel, first, lets state that we have more spaces, and here's one more idea for punctuation: 
- Following the simetric logic, also in upper center base keys 
- Adding (« ») in third layer on 5 and 6, following engram pattern

![image](https://user-images.githubusercontent.com/12291843/130324883-271b8991-f6ad-40ad-838e-6b2fd8d95ac7.png)

As you see, we have 2 more keys at the right side of the keyboard, so we still have some freedom to put more characters.

**Reply by Lobo-Feroz** [reply]
Hey @NickG13 thanks for the feedback!

> Adding (« ») in third layer on 5 and 6, following engram pattern

Agree. That was the plan, to have them on those keys.

> we have 2 more keys at the right side of the keyboard

Yep, though they are less comfortable and also one of them is displaced between ANSI and ISO, so I thought I'd leave them for later. Engram-en has there # and @ , which IMHO is a good choice.

> Following the simetric logic, also in upper center base keys

I don't know if having the parenthesis in the base layer is the best idea. According to the frequency sheet, the hyphen is more frequent than the parenthesis. We could have them in the shift layer, which is also where engram-en has them.

I was going to point out that the underscore has also more frequency than the opening parenthesis, so I wasn't eager to move it to the AltGr layer, but then I thought something: why does the underscore have more frequency than the opening parenthesis?

The underscore is used to represent spaces in names or variables where you can put a space, like an username, e.g. "Lobo_Feroz".

But in regular spanish language it's not used very much, if at all. Just like in english. See in these discussions how many times we do use hyphen or parenthesis, and we almost never use the underscore.

In fact, in Leipzig corpus there are no underscores, has a count of zero. But in Gutenberg corpus it has such as high count as to make the average rank above the quotes, the parenthesis, the exclamation sign... it doesn't make sense to me. Something is fishy with the presence of the underscore in Gutemberg. Heads up @iandoug. I'll investigate this.

Edit: I found something. Gutemberg uses underscores to represent italics. Just wrote about it in the Corpus discussion thread. 

So let's move the underscore to the AltGr layer (if we add the dash here, it'll have to be in the Shift+AltGr). We can move the ' too, very low frequency. And let's place the parenthesis in the shift layer.

![layout - signs - comma - accent - interrogation - hypen - semicolon - parenthesis](https://user-images.githubusercontent.com/10242383/130328526-5696eb9f-4d00-44b7-8e60-57f718a2e8f5.png)

I'm adding here a tfl keyboard with just these signs for ease of visualization:

[Engram-ES - V0 signs only - 20200821.tlf.zip](https://github.com/binarybottle/engram-es/files/7025820/Engram-ES.-.V0.signs.only.-.20200821.tlf.zip)

**Reply by NickG13** [reply]
There's also * that may be used to represent bold

**Reply by Lobo-Feroz** [reply]
> There's also * that may be used to represent bold

@NickG13 , I just made a new comment about that in the Corpus discussion, here: https://github.com/binarybottle/engram-es/discussions/21#discussioncomment-1216806
